### DISCUSSION TITLE
[ ALG101 ]精準度問題

### DISCUSSION THREAD

**Discussion by Glenn8119** [Lidemy/forum#67] [前端相關問題] [answered]
想請問大家，在ALG101的unit3.2中提到JS不能處理太大的數字否則會失去精準度， 介於負的 2^31 到 2^31-1，但是又有另外一個標準 Number.MAX_SAFE_INTEGER，這兩個標準的值不太一樣，想問一下差別在哪裡呢

![image](https://user-images.githubusercontent.com/71224644/115817681-dfca6900-a42d-11eb-89d1-4750eb545453.png)

**Comment by outshaker** [comment, marked as answer]
int 一般來說是 4 bytes = 32 bits ，所以最多只能表達 2^32 次方，分配給正負號就變成 2^31 次方。
而 js Number內部採用 double型態儲存，裡面可以儲存實際數值的長度為 52 bit ，詳細可以查詢 IEEE754 標準

**Reply by aszx87410** [reply]
這邊講到的 int 指的是一般程式語言中的 int 型態，但 JS 沒有這個形態，就像樓上說的 JS 只有 Number 這個形態然後都用 double 來儲存，能儲存到的最大的正整數就是 Number.MAX_SAFE_INTEGER，超過這範圍雖然一樣可以存但會不精準

``` js
console.log(Number.MAX_SAFE_INTEGER) // 9007199254740991
console.log(Number.MAX_SAFE_INTEGER+1) // 9007199254740992
console.log(Number.MAX_SAFE_INTEGER+2) // 9007199254740992
console.log(Number.MAX_SAFE_INTEGER+1 === Number.MAX_SAFE_INTEGER+2) // true
```

**Reply by Glenn8119** [reply]
原來是儲存型態不一樣的意思, 謝謝兩位解惑!
